Migrate `pkg/proxy/util` to structured logging

CIPHERTron opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/wg structured-logging 
/area logging
/priority important-longterm
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Migrate kube-proxy to structured logs

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Reference #104872 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
migrate pkg/proxy to structured logs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

CIPHERTron committed:
Migrate to Structured Logs in `pkg/proxy/util`

serathius commented:
This PR is not correctly tagged for migration, please read instructions in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#what-to-include-in-the-pull-request

CIPHERTron commented:
/wg structured-logging
/area logging
/priority important-longterm
/cc @kubernetes/wg-structured-logging-reviews

CIPHERTron commented:
> This PR is not correctly tagged for migration, please read instructions in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#what-to-include-in-the-pull-request

Sorry for the inconvenience @serathius. Is it okay now?

bowei commented:
/ok-to-test

/lgtm
/approve

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -367,7 +367,7 @@ func EnsureSysctl(sysctl utilsysctl.Interface, name string, newVal int) error {
 			return fmt.Errorf("can't set sysctl %s to %d: %v", name, newVal, err)
 		}
-		klog.V(1).Infof("Changed sysctl %q: %d -> %d", name, oldVal, newVal)
+		klog.V(1).InfoS("Changed sysctl", "name", name, "oldValue", oldVal, "newValue", newVal)
```
can we use key names "before" and "after"

CIPHERTron replied:
Sure.

serathius commented:
/hold
To address thockin comments

CIPHERTron commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -81,12 +81,12 @@ func (d *detectLocalByCIDR) IsImplemented() bool {
 func (d *detectLocalByCIDR) JumpIfLocal(args []string, toChain string) []string {
 	line := append(args, "-s", d.cidr, "-j", toChain)
-	klog.V(4).Info("[DetectLocalByCIDR (", d.cidr, ")", " Jump Local: ", line)
+	klog.V(4).InfoS("Detect Local By CIDR", "DetectLocalByCIDR", d.cidr, "jumpLocal", line)
```
It should be `d.cidr` I guess because there's no declaration that goes by the name `cidr` 🤔

thockin replied:
the key name does not have to match code literally.  This logline is saying "We are using a CIDR for this" and "the cidr is..."

So "cidr" is appropriate for the key - it  is unambiguous.

CIPHERTron replied:
My bad. I was thinking about the value instead of the key 😅

CIPHERTron commented on pkg/proxy/util/utils.go:
```diff
@@ -176,12 +176,12 @@ func GetLocalAddrSet() netutils.IPSet {
 	// if ClusterIP is "None" or empty, skip proxying
 	if !helper.IsServiceIPSet(service) {
-		klog.V(3).Infof("Skipping service %s in namespace %s due to clusterIP = %q", service.Name, service.Namespace, service.Spec.ClusterIP)
+		klog.V(3).InfoS("Skipping service in namespace due to cluster IP", "service", service.Name, "namespace", service.Namespace, "clusterIP", service.Spec.ClusterIP)
```
Yes, correct. Btw, did you mean `klog.KObj(service)`?
Also, don't we have to log the `namespace` key?

serathius commented:
/wg structured-logging 
/area logging
/priority important-soon
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

serathius commented:
/triage accepted

CIPHERTron committed:
Minor fixes

CIPHERTron commented:
Heyy @serathius @thockin @bowei I've made the requested changes except for the two conversations that are unresolved i.e. [this](https://github.com/kubernetes/kubernetes/pull/104908#discussion_r706356124) and [this](https://github.com/kubernetes/kubernetes/pull/104908#discussion_r706356205).

Should I squash merge the commits from my side? Thanks!

shivanshu1333 commented:
Kindly unlink the issue by removing `Fixes X` as this will close the original Issue, instead you can give reference to the issue.

thockin commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -81,12 +81,12 @@ func (d *detectLocalByCIDR) IsImplemented() bool {
 func (d *detectLocalByCIDR) JumpIfNotLocal(args []string, toChain string) []string {
 	line := append(args, "!", "-s", d.cidr, "-j", toChain)
-	klog.V(4).Info("[DetectLocalByCIDR (", d.cidr, ")]", " Jump Not Local: ", line)
+	klog.V(4).InfoS("Detect Local By CIDR", "DetectLocalByCIDR", d.cidr, "jumpNotLocal", line)
```
same

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -176,12 +176,12 @@ func GetLocalAddrSet() netutils.IPSet {
 	// if ClusterIP is "None" or empty, skip proxying
 	if !helper.IsServiceIPSet(service) {
-		klog.V(3).Infof("Skipping service %s in namespace %s due to clusterIP = %q", service.Name, service.Namespace, service.Spec.ClusterIP)
+		klog.V(3).InfoS("Skipping service due to cluster IP", "service", klog.KObj(service), "namespace", service.Namespace, "clusterIP", service.Spec.ClusterIP)
```
don't need "namespace" arg any more?

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -176,12 +176,12 @@ func GetLocalAddrSet() netutils.IPSet {
 	// Even if ClusterIP is set, ServiceTypeExternalName services don't get proxied
 	if service.Spec.Type == v1.ServiceTypeExternalName {
-		klog.V(3).Infof("Skipping service %s in namespace %s due to Type=ExternalName", service.Name, service.Namespace)
+		klog.V(3).InfoS("Skipping service in namespace due to Type=ExternalName", "service", klog.KObj(service), "namespace", service.Namespace)
```
same

CIPHERTron committed:
change key to cidr and remove namespace arg

CIPHERTron commented:
In the latest commits, I've changed the key to `cidr` in `pkg/proxy/util/iptables/traffic.go` and removed `namespace` arg in `pkg/proxy/util/utils.go`. @thockin thanks for your review, are there any more changes that needs to be done?

yangjunmyfm192085 commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -81,12 +81,12 @@ func (d *detectLocalByCIDR) IsImplemented() bool {
 func (d *detectLocalByCIDR) JumpIfLocal(args []string, toChain string) []string {
 	line := append(args, "-s", d.cidr, "-j", toChain)
-	klog.V(4).Info("[DetectLocalByCIDR (", d.cidr, ")", " Jump Local: ", line)
+	klog.V(4).InfoS("Detect Local By CIDR", "cidr", d.cidr, "jumpLocal", line)
```
```suggestion
	klog.V(4).InfoS("Detect Local By CIDR", "CIDR", d.cidr, "jumpLocal", line)
```

yangjunmyfm192085 commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -81,12 +81,12 @@ func (d *detectLocalByCIDR) IsImplemented() bool {
 func (d *detectLocalByCIDR) JumpIfNotLocal(args []string, toChain string) []string {
 	line := append(args, "!", "-s", d.cidr, "-j", toChain)
-	klog.V(4).Info("[DetectLocalByCIDR (", d.cidr, ")]", " Jump Not Local: ", line)
+	klog.V(4).InfoS("Detect Local By CIDR", "cidr", d.cidr, "jumpNotLocal", line)
```
```suggestion
	klog.V(4).InfoS("Detect Local By CIDR", "CIDR", d.cidr, "jumpNotLocal", line)
```

yangjunmyfm192085 commented on pkg/proxy/util/utils.go:
```diff
@@ -254,7 +254,7 @@ func GetNodeAddresses(cidrs []string, nw NetworkInterfacer) (sets.String, error)
 func LogAndEmitIncorrectIPVersionEvent(recorder events.EventRecorder, fieldName, fieldValue, svcNamespace, svcName string, svcUID types.UID) {
 	errMsg := fmt.Sprintf("%s in %s has incorrect IP version", fieldValue, fieldName)
-	klog.Errorf("%s (service %s/%s).", errMsg, svcNamespace, svcName)
+	klog.ErrorS(nil, "Incorrect IP version", "service", klog.KRef(svcName, svcNamespace), "field", fieldName, "value", fieldValue)
```
klog.KRef(podNamespace, podName)

CIPHERTron replied:
@thockin suggested it should be `klog.KRef(svcName, svcNamespace)` and if we see, we have to log service name and namespace and not pod name and namespace. I might be wrong though. // @yangjunmyfm192085

yangjunmyfm192085 replied:
> @thockin suggested it should be `klog.KRef(svcName, svcNamespace)` and if we see, we have to log service name and namespace and not pod name and namespace. I might be wrong though. // @yangjunmyfm192085

Sorry， my mistake, my mean, it's  `klog.KRef(svcNamespace,svcName)`

yangjunmyfm192085 replied:
You can change to
`klog.ErrorS(nil, "Incorrect IP version", "service", klog.KRef(svcNamespace, svcName,), "field", fieldName, "value", fieldValue)`

CIPHERTron replied:
Done @yangjunmyfm192085 🚀

yangjunmyfm192085 replied:
thanks

yangjunmyfm192085 commented on pkg/proxy/util/utils.go:
```diff
@@ -274,7 +274,7 @@ func MapIPsByIPFamily(ipStrings []string) map[v1.IPFamily][]string {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], ip)
 		} else {
-			klog.Errorf("Skipping invalid IP: %s", ip)
+			klog.ErrorS(nil, "Skipping invalid IP", "ip", ip)
```
```suggestion
			klog.ErrorS(nil, "Skipping invalid IP", "IP", ip)
```

yangjunmyfm192085 commented on pkg/proxy/util/utils.go:
```diff
@@ -288,7 +288,7 @@ func MapCIDRsByIPFamily(cidrStrings []string) map[v1.IPFamily][]string {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], cidr)
 		} else {
-			klog.Errorf("Skipping invalid cidr: %s", cidr)
+			klog.ErrorS(nil, "Skipping invalid CIDR", "cidr", cidr)
```
```suggestion
			klog.ErrorS(nil, "Skipping invalid CIDR", "CIDR", cidr)
```

yangjunmyfm192085 commented:
Please refer to `https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md`

CIPHERTron committed:
Update key from cidr to CIDR

Co-authored-by: JUN YANG <69306452+yangjunmyfm192085@users.noreply.github.com>

CIPHERTron committed:
Update key cidr to CIDR

Co-authored-by: JUN YANG <69306452+yangjunmyfm192085@users.noreply.github.com>

CIPHERTron committed:
Update key ip to IP

Co-authored-by: JUN YANG <69306452+yangjunmyfm192085@users.noreply.github.com>

CIPHERTron committed:
Update key ip to IP

Co-authored-by: JUN YANG <69306452+yangjunmyfm192085@users.noreply.github.com>

yangjunmyfm192085 commented:
/ok-to-test

shivanshu1333 commented on pkg/proxy/util/endpoints.go:
```diff
@@ -53,12 +53,12 @@ func PortPart(s string) (int, error) {
 	portNumber, err := strconv.Atoi(port)
 	if err != nil {
-		klog.Errorf("Error parsing '%s': %v", port, err)
+		klog.ErrorS(err, "failed to parse port", "input", port)
```
Shouldn't we use "port" instead of "input"?
cc: @thockin @serathius

mengjiao-liu replied:
/hold

Wait for the commit content to be resolved.

CIPHERTron committed:
Interchange svcNamespace and svcName

CIPHERTron commented on pkg/proxy/util/utils.go:
```diff
@@ -274,7 +274,7 @@ func MapIPsByIPFamily(ipStrings []string) map[v1.IPFamily][]string {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], ip)
 		} else {
-			klog.Errorf("Skipping invalid IP: %s", ip)
+			klog.ErrorS(nil, "Skipping invalid IP", "ip", ip)
```
I've already changed this. Not sure why it is not showing here 😅

CIPHERTron commented:
Heyy folks, can we wrap up this PR? Thanks!
// @serathius @thockin @yangjunmyfm192085

CIPHERTron commented:
> > Heyy folks, can we wrap up this PR? Thanks!
> > // @serathius @thockin @yangjunmyfm192085
> 
> LGTM to me, except
> 
> ```
> Additionally we can improve messages to comply with good practices:
> 
> Start from a capital letter.
> ```
> 
> refer to `https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md`
> we need @serathius and @thockin review

Oh yeah, now they are kinda mixed up. Some start with a capital and some with a small.
Should I add a final commit changing the first letter of each message with a capital letter?

yangjunmyfm192085 commented:
> > > Heyy folks, can we wrap up this PR? Thanks!
> > > // @serathius @thockin @yangjunmyfm192085
> > 
> > 
> > LGTM to me, except
> > ```
> > Additionally we can improve messages to comply with good practices:
> > 
> > Start from a capital letter.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > refer to `https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md`
> > we need @serathius and @thockin review
> 
> Oh yeah, now they are kinda mixed up. Some start with a capital and some with a small.
> Should I add a final commit changing the first letter of each message with a capital letter?

I think that would be better,

CIPHERTron committed:
Change first letter of all messages to capital

CIPHERTron commented:
Done @yangjunmyfm192085 🚀

yangjunmyfm192085 reviewed (commented):
/cc @serathius @thockin

mengjiao-liu commented:
/label tide/merge-method-squash
/lgtm

shivanshu1333 commented on pkg/proxy/util/endpoints.go:
```diff
@@ -36,13 +36,13 @@ func IPPart(s string) string {
 	host, _, err := net.SplitHostPort(s)
 	if err != nil {
-		klog.Errorf("Error parsing '%s': %v", s, err)
+		klog.ErrorS(err, "Failed to parse host-port", "input", s)
```
change "input" to "hostport"

mengjiao-liu replied:
+1
I thought it would be easier to write parameters "hostport "  directly instead of using "input".

CIPHERTron replied:
Okay, cool
 I'll make changes accordingly 🚀

shivanshu1333 commented on pkg/proxy/util/endpoints.go:
```diff
@@ -36,13 +36,13 @@ func IPPart(s string) string {
 	ip := netutils.ParseIPSloppy(host)
 	if ip == nil {
-		klog.Errorf("invalid IP part '%s'", host)
+		klog.ErrorS(nil, "Failed to parse IP", "input", host)
```
change "input" to "host"

shivanshu1333 commented on pkg/proxy/util/endpoints.go:
```diff
@@ -53,12 +53,12 @@ func PortPart(s string) (int, error) {
 	_, port, err := net.SplitHostPort(s)
 	if err != nil {
-		klog.Errorf("Error parsing '%s': %v", s, err)
+		klog.ErrorS(err, "Failed to parse host-port", "input", s)
```
change "input" to "hostport"

shivanshu1333 commented on pkg/proxy/util/endpoints.go:
```diff
@@ -53,12 +53,12 @@ func PortPart(s string) (int, error) {
 	portNumber, err := strconv.Atoi(port)
 	if err != nil {
-		klog.Errorf("Error parsing '%s': %v", port, err)
+		klog.ErrorS(err, "Failed to parse port", "input", port)
```
change "input" to "port"

shivanshu1333 commented:
\hold

shivanshu1333 commented:
/hold

mengjiao-liu commented:
> using port and hostport instead of input makes more sense, in the other migration PRs I've seen the same practice.

/hold
Wait for the commit content to be resolved.

CIPHERTron commented:
> using port and hostport instead of input makes more sense, in the other migration PRs I've seen the same practice.
> /hold

Should I change them @shivanshu1333?

CIPHERTron committed:
Change key names in endpoints.go

CIPHERTron commented:
In my latest commit, I've changed the key names in `endpoints.go` as requested by @shivanshu1333 👍🏻

shivanshu1333 commented:
@CIPHERTron can you please fix the issue in your latest commit? Thanks!

CIPHERTron commented:
> @CIPHERTron can you please fix the issue in your latest commit? Thanks!

I'm not quite sure what's causing the tests to fail. Could you please help me out?

shivanshu1333 commented:
> This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
see https://github.com/kubernetes/kubernetes/commit/39a69521438526b37d57f4e5dad53ea14b662298

CIPHERTron commented:
> > This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
> 
> see [39a6952](https://github.com/kubernetes/kubernetes/commit/39a69521438526b37d57f4e5dad53ea14b662298)

Isn't it supposed to be a commit from a branch on my fork?
Like, for this PR, every commit belongs to `CIPHERTron/kubernetes` fork's `structured-logs` branch
Or maybe I'm not getting what the error means. Pretty new to it actually 😅

karuppiah7890 commented:
@CIPHERTron Nothing is wrong with your commits. Your commits do belong to the branch of your fork. Turns out Kubernetes bot is not brainy enough to use the right link for the commits in this PR in it's [comment](https://github.com/kubernetes/kubernetes/pull/104908#issuecomment-920805457) OR there is some issue with GitHub linking issues, which seems to be the case with some experimentation. The right link for the commit is - [this](https://github.com/kubernetes/kubernetes/pull/104908/commits/39a69521438526b37d57f4e5dad53ea14b662298) or [this](https://github.com/CIPHERTron/kubernetes/commit/39a69521438526b37d57f4e5dad53ea14b662298). If you simply try to paste the kubernetes/kubernetes repo based commit link as is, that is this link `https://github.com/kubernetes/kubernetes/pull/104908/commits/39a69521438526b37d57f4e5dad53ea14b662298` in a GitHub comment, it gets rendered like this - https://github.com/kubernetes/kubernetes/pull/104908/commits/39a69521438526b37d57f4e5dad53ea14b662298 and when you hover over it or click it, it leads to the wrong link because GitHub seems to have some issue. Kubernetes bot can maybe just use different way to link to the commits I guess. Or we gotta report to GitHub

So, that's the whole confusion I think. It's no way related to the test failure. The test failure is here

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/104908/pull-kubernetes-unit/1438452621122736128

```bash


k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure: TestCreateBlobDisk expand_less | 0s
-- | --
=== RUN   TestCreateBlobDisk panic: test timed out after 3m0s  goroutine 232 [running]: testing.(*M).startAlarm.func1() 	/usr/local/go/src/testing/testing.go:1788 +0xbb created by time.goFunc 	/usr/local/go/src/time/sleep.go:180 +0x4a  goroutine 1 [chan receive]: testing.(*T).Run(0xc0002921a0, {0x251cb43, 0x12}, 0x25e6c40)

```

as mentioned in bot's [comment](https://github.com/kubernetes/kubernetes/pull/104908#issuecomment-920805457) with the link test as `link`

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -288,7 +288,7 @@ func MapCIDRsByIPFamily(cidrStrings []string) map[v1.IPFamily][]string {
 			ipFamilyMap[ipFamily] = append(ipFamilyMap[ipFamily], cidr)
 		} else {
-			klog.Errorf("Skipping invalid cidr: %s", cidr)
+			klog.ErrorS(nil, "Skipping invalid CIDR", "CIDR", cidr)
```
"cidr"

thockin reviewed (commented):
sorry to nit-pick.  (not really sorry).

In general I think we want key names to be "lowerBumpyCaps" - "cidr" not "CIDR", etc

shivanshu1333 commented:
@CIPHERTron can you please take a look at the review comments,
Thanks!

CIPHERTron commented:
> @CIPHERTron can you please take a look at the review comments,
> Thanks!

Yeah, sorry about that. Was caught up with something else today!

CIPHERTron committed:
Change all keynames to lower bumby caps convention

CIPHERTron commented:
Heyy @thockin, I've addressed all your suggestions. Could you please do a final check and say if everything's good to go or it needs some more changes? Thanks a lot!
/cc @serathius

shivanshu1333 commented:
/test pull-kubernetes-integration

shivanshu1333 commented:
/test pull-kubernetes-integration

CIPHERTron commented:
Heyy folks, do I have to squash merge it from my side? Thanks!
/cc @serathius @thockin @shivanshu1333

serathius commented:
/lgtm

serathius commented:
/unhold
